File: clients/typescript/src/action_server.ts
import * as readline from 'node:readline';
import { ActionRequest } from './dtos';
import { clientSetup, AmenoException } from './shared';

const ON_STARTUP_SUCCESS = '[AmenoLink.StartupSuccess]';
const ON_ACTION_SUCCESS = '[AmenoLink.ActionSuccess]';
const ON_ACTION_ERROR = '[AmenoLink.ActionError]';
const ON_ACTION_LOGGED = '[AmenoLink.ActionLog]';

export type ActionHandler<T, R> = (input: T) => Promise<R> | R;

export interface ActionRoute {
    route: string;
    handler: ActionHandler<any, any>;
}

export class ActionContext {
    public readonly request: ActionRequest<unknown>;

    constructor(request: ActionRequest<unknown>) {
        this.request = request;
    }

    public log(message: string): void {
        sendMessage(ON_ACTION_LOGGED, message);
    }
}

export class ActionRouter {
    private readonly routes: ActionRoute[] = [];

    public add<T, R>(route: string, handler: ActionHandler<T, R>): void {
        this.routes.push({ route, handler });
    }

    private async execute(request: ActionRequest<any>): Promise<string> {
        for (const route of this.routes) {
            if (route.route === request.route) {
                const rawResult = await route.handler(request.payload);
                return this.formatResult(rawResult);
            }
        }
        throw new AmenoException(`Rota '${request.route}' não encontrada`);
    }

    private formatResult(rawResult: any): string {
        return JSON.stringify(rawResult ?? null);
    }

    public serve(): void {
        sendMessage(ON_STARTUP_SUCCESS, clientSetup.appName);

        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            terminal: false
        });

        rl.on('line', async (line: string) => {
            const trimmedLine = line.trim();
            if (!trimmedLine)
                return;

            try {
                const decodedJson = Buffer.from(trimmedLine, 'base64').toString('utf-8');
                const request: ActionRequest<any> = JSON.parse(decodedJson);

                currentAction = new ActionContext(request);
                const result = await this.execute(request);
                sendMessage(ON_ACTION_SUCCESS, result);
            } catch (error: any) {
                sendMessage(ON_ACTION_ERROR, error?.message ?? String(error));
            } finally {
                currentAction = null;
            }
        });

        rl.on('error', (error: any) => {
            sendMessage(ON_ACTION_ERROR, error?.message ?? String(error));
        });
    }
}

export const actions = new ActionRouter();
let currentAction: ActionContext | null = null;

export function actionContext(): ActionContext {
    if (!currentAction)
        throw new Error('Nenhuma ação está em execução no momento.');

    return currentAction;
}

function sendMessage(prefix: string, message: string): void {
    const payload = Buffer.from(message, 'utf-8').toString('base64');
    process.stdout.write(`${prefix}${payload}\n`);
}

File: clients/csharp/AmenoLink/ActionServer.cs
using System.Text.Json;

namespace AmenoLink;

public class ActionContext(ActionRequest<object?> request) : IActionContext
{
    public ActionRequest<object?> Request { get; } = request;

    public void Log(string message)
    {
        ActionServer.SendMessage(ActionServer.OnActionLogged, message);
    }
}

public class ActionRouter : IActionRouter
{
    private record RouteEntry(string Route, Func<object?, Task<string>> Handler);

    private readonly List<RouteEntry> routes = [];

    public void Add<T, R>(string route, Func<T, Task<R>> handler)
    {
        routes.Add(new RouteEntry(route, async rawPayload =>
        {
            T? input = DeserializePayload<T>(rawPayload);
            R result = await handler(input!);
            return JsonSerializer.Serialize(result, AmenoLinkClient.JsonOptions);
        }));
    }

    public void Add<T, R>(string route, Func<T, R> handler)
    {
        routes.Add(new RouteEntry(route, rawPayload =>
        {
            T? input = DeserializePayload<T>(rawPayload);
            R result = handler(input!);
            string json = JsonSerializer.Serialize(result, AmenoLinkClient.JsonOptions);
            return Task.FromResult(json);
        }));
    }

    public void Serve()
    {
        ActionServer.Serve(this);
    }

    internal async Task<string> Execute(ActionRequest<object?> request)
    {
        foreach (var entry in routes)
        {
            if (entry.Route == request.Route)
                return await entry.Handler(request.Payload);
        }

        throw new AmenoException($"Rota '{request.Route}' não encontrada");
    }

    private static T? DeserializePayload<T>(object? rawPayload)
    {
        if (rawPayload is null)
            return default;

        if (rawPayload is T typed)
            return typed;

        if (rawPayload is JsonElement jsonElement)
            return JsonSerializer.Deserialize<T>(jsonElement.GetRawText(), AmenoLinkClient.JsonOptions);

        string json = JsonSerializer.Serialize(rawPayload, AmenoLinkClient.JsonOptions);
        return JsonSerializer.Deserialize<T>(json, AmenoLinkClient.JsonOptions);
    }
}

public static class ActionServer
{
    public const string OnStartupSuccess = "[AmenoLink.StartupSuccess]";
    public const string OnActionSuccess = "[AmenoLink.ActionSuccess]";
    public const string OnActionError = "[AmenoLink.ActionError]";
    public const string OnActionLogged = "[AmenoLink.ActionLog]";

    private static readonly AsyncLocal<ActionContext?> CurrentActionContext = new();

    public static ActionContext ActionContext
    {
        get
        {
            var current = CurrentActionContext.Value;
            if (current == null)
                throw new InvalidOperationException("Nenhuma ação está em execução no momento.");

            return current;
        }
    }

    public static void Serve(ActionRouter router)
    {
        SendMessage(OnStartupSuccess, AmenoLinkClient.Settings.AppName);

        string? line;
        while ((line = Console.ReadLine()) != null)
        {
            string trimmedLine = line.Trim();
            if (string.IsNullOrEmpty(trimmedLine))
                continue;

            try
            {
                byte[] decodedBytes = Convert.FromBase64String(trimmedLine);
                string decodedJson = System.Text.Encoding.UTF8.GetString(decodedBytes);
                var request = JsonSerializer.Deserialize<ActionRequest<object?>>(decodedJson, AmenoLinkClient.JsonOptions);

                if (request == null)
                    continue;

                var context = new ActionContext(request);
                CurrentActionContext.Value = context;

                string resultJson = router.Execute(request).GetAwaiter().GetResult();
                SendMessage(OnActionSuccess, resultJson);
            }
            catch (Exception exception)
            {
                SendMessage(OnActionError, exception.Message);
            }
            finally
            {
                CurrentActionContext.Value = null;
            }
        }
    }

    internal static void SendMessage(string prefix, string message)
    {
        byte[] payloadBytes = System.Text.Encoding.UTF8.GetBytes(message);
        string payload = Convert.ToBase64String(payloadBytes);
        Console.WriteLine($"{prefix}{payload}");
    }
}
